drivers/Watchdog.h

```diff
@@ -0,0 +1,174 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2017 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef MBED_WATCHDOG_H
+#define MBED_WATCHDOG_H
+
+#ifdef DEVICE_WATCHDOG
+
+#include "watchdog_api.h"
+
+#include <cstdio>
+
+
+namespace mbed {
+
+/** A system timer that will reset the system in the case of system failures or
+ *  malfunctions.
+ *
+ * Example:
+ * @code
+ *
+ * Watchdog watchdog = Watchdog();
+ * watchdog.set_timeout(2000);
+ * watchdog.start();
+ *
+ * while (true) {
+ *    watchdog.kick();
+ *
+ *    // Application code
+ * }
+ * @endcode
+ *
+ */
+class Watchdog
+{
+private:
+  uint32_t timeout_;
+  uint32_t window_;
+  bool enable_sleep_;
+
+public:
+    /** Create a watchdog timer with no configured parameters
+     */
+    Watchdog() : timeout_(0), window_(0), enable_sleep_(true) {}
+
+
+    /** Create a Watchdog timer with specified parameters
+     *
+     *  @param timeout Timeout of the watchdog in milliseconds
+     *  @param window  Time period of the window of the watchdog [default: 0]
+
+     *  @note
```

Comment:
What if the windowed mode is not supported? Should we have empty default constructor and 'init' function that can return error?

Reply:
There's currently no function to determine if window mode is supported which makes it difficult, it needs to be added to the HAL API and maybe as a DEVICE_WATCHDOG_WINDOW so it can be returned correctly

Reply:
Why not remove all the parameters from watchdog driver constructor. And move all the logic to start which would call `hal_watchdog_init` which returns error in case the window mode is not supported.

Reply:
Yeah, that makes more sense